From 8f7122908220b585f9ee8ec8470b9a6228a3bb79 Mon Sep 17 00:00:00 2001
From: NLObP <38990417+NL0bP@users.noreply.github.com>
Date: Tue, 13 Aug 2024 21:22:29 +0300
Subject: [PATCH] [Game] Fixed melee and archery animations: - since the melee
 animation is Id=0, we had to make a list of animations that are displayed
 sequentially; - added a delay to Cast of 300 ms;

---
 AAEmu.Commons/Utils/Rand.cs                   |  13 ++
 .../Core/Packets/C2G/CSStartSkillPacket.cs    |  34 ++-
 .../Core/Packets/G2C/SCSkillFiredPacket.cs    | 216 ++++++++++++++++--
 .../Effects/SpecialEffects/AutoAttack.cs      |   2 +
 AAEmu.Game/Models/Game/Skills/Skill.cs        |  12 +-
 5 files changed, 241 insertions(+), 36 deletions(-)

diff --git a/AAEmu.Commons/Utils/Rand.cs b/AAEmu.Commons/Utils/Rand.cs
index d53fad850e..5671d49ab1 100644
--- a/AAEmu.Commons/Utils/Rand.cs
+++ b/AAEmu.Commons/Utils/Rand.cs
@@ -1,4 +1,5 @@
 ﻿using System;
+using System.Collections.Generic;
 
 namespace AAEmu.Commons.Utils;
 
@@ -62,4 +63,16 @@ public static float Next(float minValue, float maxValue)
             return _random.NextSingle(true) * (maxValue - minValue) + minValue;
         }
     }
+
+    public static int GetRandomElement(List<int> list)
+    {
+        // Создаем экземпляр класса Random
+        var random = new Random();
+
+        // Получаем случайный индекс в пределах длины списка
+        var randomIndex = random.Next(list.Count);
+
+        // Возвращаем элемент по случайному индексу
+        return list[randomIndex];
+    }
 }
diff --git a/AAEmu.Game/Core/Packets/C2G/CSStartSkillPacket.cs b/AAEmu.Game/Core/Packets/C2G/CSStartSkillPacket.cs
index 3512588fd2..9e29e8e83f 100644
--- a/AAEmu.Game/Core/Packets/C2G/CSStartSkillPacket.cs
+++ b/AAEmu.Game/Core/Packets/C2G/CSStartSkillPacket.cs
@@ -4,11 +4,9 @@
 
 using AAEmu.Commons.Network;
 using AAEmu.Game.Core.Managers;
-using AAEmu.Game.Core.Managers.Id;
 using AAEmu.Game.Core.Managers.World;
 using AAEmu.Game.Core.Network.Game;
 using AAEmu.Game.Core.Packets.G2C;
-using AAEmu.Game.Models.Game;
 using AAEmu.Game.Models.Game.Char;
 using AAEmu.Game.Models.Game.Items.Templates;
 using AAEmu.Game.Models.Game.Skills;
@@ -29,7 +27,7 @@ public override void Read(PacketStream stream)
         using var source = new CancellationTokenSource();
         var t = Task.Run(async delegate
         {
-            await Task.Delay(TimeSpan.FromMilliseconds(100), source.Token);
+            await Task.Delay(TimeSpan.FromMilliseconds(150), source.Token);
             return 0;
         });
         try
@@ -58,7 +56,7 @@ public override void Read(PacketStream stream)
         if (flagType > 0) skillObject.Read(stream);
 
         Logger.Info($"StartSkill: Id {skillId}, flag {flag}, caster={skillCaster.ObjId}, target={skillCastTarget.ObjId}");
-        
+
         var skillResult = SkillResult.Success;
         var skillResultErrorValue = 0u;
         Skill skill = null;
@@ -103,28 +101,28 @@ public override void Read(PacketStream stream)
             // Execute the rider/operator skill as the player using either target or self
             skillResult = Connection.ActiveChar.UseSkill(mountAttachedSkill, riderTarget ?? Connection.ActiveChar);
         }
-        else if (Connection.ActiveChar.IsAutoAttack && skillId == Connection.ActiveChar.AutoAttackTask?.Skill?.Template?.Id)
-        {
-            // Same as already executing auto-skill, just send the success result.
-            skill = Connection.ActiveChar.AutoAttackTask.Skill;
-            skillResult = SkillResult.Success;
-        }
+        //else if (Connection.ActiveChar.IsAutoAttack && skillId == Connection.ActiveChar.AutoAttackTask?.Skill?.Template?.Id)
+        //{
+        //    // Same as already executing auto-skill, just send the success result.
+        //    skill = Connection.ActiveChar.AutoAttackTask.Skill;
+        //    skillResult = SkillResult.Success;
+        //}
         else if (SkillManager.Instance.IsDefaultSkill(skillId) || SkillManager.Instance.IsCommonSkill(skillId) && !(skillCaster is SkillItem))
         {
             // Is it a common skill?
             skill = new Skill(SkillManager.Instance.GetSkillTemplate(skillId)); // TODO: переделать / rewrite ...
             skillResult = skill.Use(Connection.ActiveChar, skillCaster, skillCastTarget, skillObject, false, out skillResultErrorValue);
-            if ((skillResult == SkillResult.Success) && (skillId < 5000) && (skillCaster.ObjId == Connection.ActiveChar.ObjId))
-            {
-                // All basic combat skills are below ID 5000, only 2 (melee),3 (offhand) and 4 (ranged) exist, next actual skill used is 5001
-                Connection.ActiveChar.IsAutoAttack = true;
-                Connection.ActiveChar.StartAutoSkill(skill);
-            }
+            //if ((skillResult == SkillResult.Success) && (skillId < 5000) && (skillCaster.ObjId == Connection.ActiveChar.ObjId))
+            //{
+            //    // All basic combat skills are below ID 5000, only 2 (melee),3 (offhand) and 4 (ranged) exist, next actual skill used is 5001
+            //    Connection.ActiveChar.IsAutoAttack = true;
+            //    Connection.ActiveChar.StartAutoSkill(skill);
+            //}
         }
         else if (skillCaster is SkillItem si)
         {
             // A skill triggered by a item
-            var player = Connection.ActiveChar; 
+            var player = Connection.ActiveChar;
             // var item = player.Inventory.GetItemById(si.ItemId);
             // добавил проверку на ItemBindType.BindOnPickup для записи портала с помощью камина в доме
             if (si.SkillSourceItem == null || skillId != si.SkillSourceItem.Template.UseSkillId && si.SkillSourceItem.Template.BindType != ItemBindType.BindOnPickup)
@@ -154,7 +152,7 @@ public override void Read(PacketStream stream)
             skill = new Skill(SkillManager.Instance.GetSkillTemplate(skillId));
             skillResult = skill.Use(Connection.ActiveChar, skillCaster, skillCastTarget, skillObject, false, out skillResultErrorValue);
         }
-        
+
         if (skillResult != SkillResult.Success)
         {
             // It actually sends a skill started packet, but not a skill fired or stopped
diff --git a/AAEmu.Game/Core/Packets/G2C/SCSkillFiredPacket.cs b/AAEmu.Game/Core/Packets/G2C/SCSkillFiredPacket.cs
index 7af1c04e3b..484ad2d422 100644
--- a/AAEmu.Game/Core/Packets/G2C/SCSkillFiredPacket.cs
+++ b/AAEmu.Game/Core/Packets/G2C/SCSkillFiredPacket.cs
@@ -1,28 +1,150 @@
-﻿using AAEmu.Commons.Network;
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+using AAEmu.Commons.Network;
 using AAEmu.Game.Core.Network.Game;
+using AAEmu.Game.Models.Game.Char;
 using AAEmu.Game.Models.Game.Skills;
+using AAEmu.Game.Models.Game.Units;
 
 namespace AAEmu.Game.Core.Packets.G2C;
 
-#pragma warning disable IDE0052 // Remove unread private members
-
 public class SCSkillFiredPacket : GamePacket
 {
     public override PacketLogLevel LogLevel => PacketLogLevel.Trace;
 
-    private uint _id;
-    private ushort _tl;
-    private SkillCaster _caster;
-    private SkillCastTarget _target;
-    private SkillObject _skillObject;
-    private Skill _skill;
-
+    private readonly uint _id;
+    private readonly ushort _tl;
+    private readonly SkillCaster _caster;
+    private readonly SkillCastTarget _target;
+    private readonly SkillObject _skillObject;
+    private readonly Skill _skill;
     private short _effectDelay = 37;
-    private int _fireAnimId = 2;
     private bool _dist;
+    private bool _fist;
+    private bool _oneHand;
+    private bool _twoHand;
+    private bool _bow;
+
+    private readonly Character _character;
+    private readonly int _fireAnimId;
+    private static readonly Dictionary<int, int> CharacterFireAnimDataOneHand = new()
+    {
+        { 3, 45 }, { 4, 45 }, { 5, 45 }, { 6, 45 },
+        { 87, 45 }, { 88, 45 }, { 89, 45 }, { 90, 45 },
+        { 92, 45 }, { 93, 45 }, { 94, 45 }
+    };
+    private static readonly Dictionary<int, int> CharacterFireAnimDataTwoHand = new()
+    {
+        { 7, 45 }, { 95, 45 }, { 139, 45 }
+    };
+    private static readonly Dictionary<int, int> CharacterFireAnimDataBow = new()
+    {
+        { 9, 45 }
+    };
+    private static readonly Dictionary<int, int> NpcFireAnimData = new() { { 1, 50 }, { 2, 45 } };
+
+    private static Dictionary<int, int> FireAnimData;
+    private static Queue<int> FireAnimQueue;
+    private static Queue<int> NpcFireAnimQueue;
 
     public short ComputedDelay { get; set; }
 
+    public SCSkillFiredPacket(uint id, ushort tl, SkillCaster caster, SkillCastTarget target, Skill skill, SkillObject skillObject, BaseUnit baseUnit)
+        : base(SCOffsets.SCSkillFiredPacket, 5)
+    {
+        _id = id;
+        _tl = tl;
+        _caster = caster;
+        _target = target;
+        _skill = skill;
+        _skillObject = skillObject;
+        _character = baseUnit as Character;
+
+        if (_skill.Template.Id == 2)
+        {
+            if (_character is not null)
+            {
+                var mainHandItem = _character.Equipment.GetItemBySlot(15);
+                var offHandItem = _character.Equipment.GetItemBySlot(16);
+
+                _oneHand = mainHandItem != null;
+                _fist = mainHandItem == null;
+                _twoHand = offHandItem != null;
+            }
+            else
+            {
+                _oneHand = false;
+                _twoHand = false;
+                _fist = false;
+            }
+
+            _fireAnimId = GetNextAnimationId();
+        }
+
+        if (_skill.Template.Id == 2 && _character is not null)
+        {
+            Logger.Info($"SkillFired: Id={_id}:{_fireAnimId}, caster={baseUnit.ObjId}, target={_target.ObjId}");
+        }
+    }
+
+    private int GetNextAnimationId()
+    {
+        if (_character is not null)
+        {
+            if (FireAnimQueue == null || FireAnimQueue.Count == 0)
+            {
+                FireAnimData = new Dictionary<int, int>();
+                var allKeys = Enumerable.Empty<int>();
+
+                if (_oneHand)
+                {
+                    allKeys = allKeys.Concat(CharacterFireAnimDataOneHand.Keys);
+                    MergeDictionaries(FireAnimData, CharacterFireAnimDataOneHand);
+                }
+
+                if (_twoHand)
+                {
+                    allKeys = allKeys.Concat(CharacterFireAnimDataTwoHand.Keys);
+                    MergeDictionaries(FireAnimData, CharacterFireAnimDataTwoHand);
+                }
+
+                if (_fist)
+                {
+                    allKeys = NpcFireAnimData.Keys;
+                    FireAnimData = NpcFireAnimData;
+                }
+
+                // Перемешивание ключей
+                var rng = new Random();
+                allKeys = allKeys.OrderBy(x => rng.Next());
+                FireAnimQueue = new Queue<int>(allKeys);
+            }
+            // Dequeue the next animation ID
+            return FireAnimQueue.Dequeue();
+        }
+        else
+        {
+            // если не персонаж, то Npc
+            if (NpcFireAnimQueue == null || NpcFireAnimQueue.Count == 0)
+            {
+                var allKeys = NpcFireAnimData.Keys;
+                NpcFireAnimQueue = new Queue<int>(allKeys);
+            }
+            // Dequeue the next animation ID
+            return NpcFireAnimQueue.Dequeue();
+        }
+    }
+
+    private void MergeDictionaries(Dictionary<int, int> target, Dictionary<int, int> source)
+    {
+        foreach (var kvp in source)
+        {
+            target[kvp.Key] = kvp.Value;
+        }
+    }
+
     public SCSkillFiredPacket(uint id, ushort tl, SkillCaster caster, SkillCastTarget target, Skill skill, SkillObject skillObject)
         : base(SCOffsets.SCSkillFiredPacket, 5)
     {
@@ -33,6 +155,7 @@ public SCSkillFiredPacket(uint id, ushort tl, SkillCaster caster, SkillCastTarge
         _skill = skill;
         _skillObject = skillObject;
     }
+
     public SCSkillFiredPacket(uint id, ushort tl, SkillCaster caster, SkillCastTarget target, Skill skill, SkillObject skillObject, short effectDelay = 37, int fireAnimId = 2, bool dist = true)
         : base(SCOffsets.SCSkillFiredPacket, 5)
     {
@@ -56,12 +179,73 @@ public override PacketStream Write(PacketStream stream)
         stream.Write(_target);      // SkillCastTarget
         stream.Write(_skillObject); // SkillObject
 
-        stream.Write((short)(ComputedDelay / 10 + 10)); // TODO  +10 It became visible flying arrows 
-        stream.Write((short)(_skill.Template.ChannelingTime / 10 + 10));
-        stream.Write((byte)0); // f - When changed to 1 when firing an auto-casting skill, will make the little blue arrow.
-        stream.WritePisc(_id, _skill.Template.FireAnimId); // added skill type here in 3.0.3.0
-        stream.Write((byte)0); // flag
+        if (_character is not null)
+        {
+            WriteCharacterSkillData(stream);
+        }
+        else
+        {
+            WriteNpcSkillData(stream);
+        }
 
         return stream;
     }
+
+    private void WriteCharacterSkillData(PacketStream stream)
+    {
+        if (_skill.Template.Id == 2)
+        {
+            // ближняя атака
+            if (_character?.IsAutoAttack == true)
+            {
+                // автоатака
+                stream.Write((short)0);
+                stream.Write((short)0);
+                stream.Write((byte)1);  // f - When changed to 1 when firing an auto-casting skill, will make the little blue arrow.
+                stream.Write((byte)15); // c
+                stream.WritePisc(_id, 0); // added skill type here in 3.0.3.0
+                stream.Write((byte)0); // flag
+            }
+            else
+            {
+                // отдельные удары
+                stream.Write((short)FireAnimData[_fireAnimId]); // выставляем время для ближнего боя в зависимости от анимации
+                stream.Write((short)0);
+                stream.Write((byte)0); // f - When changed to 1 when firing an auto-casting skill, will make the little blue arrow.
+                stream.WritePisc(_id, _fireAnimId); // added skill type here in 3.0.3.0
+                stream.Write((byte)0); // flag
+            }
+        }
+        else
+        {
+            // дальняя атака
+            stream.Write((short)(ComputedDelay / 10 + 10)); // TODO  +10 It became visible flying arrows 
+            stream.Write((short)(_skill.Template.ChannelingTime / 10 + 10));
+            stream.Write((byte)0); // f - When changed to 1 when firing an auto-casting skill, will make the little blue arrow.
+            stream.WritePisc(_id, _skill.Template.FireAnim?.Id ?? 0); // added skill type here in 3.0.3.0
+            stream.Write((byte)0); // flag
+        }
+    }
+
+    private void WriteNpcSkillData(PacketStream stream)
+    {
+        if (_skill.Template.Id == 2)
+        {
+            // ближняя атака
+            stream.Write((short)NpcFireAnimData[_fireAnimId]); // выставляем время для ближнего боя в зависимости от анимации
+            stream.Write((short)0);
+            stream.Write((byte)0); // f - When changed to 1 when firing an auto-casting skill, will make the little blue arrow.
+            stream.WritePisc(_id, _fireAnimId); // added skill type here in 3.0.3.0
+            stream.Write((byte)0); // flag
+        }
+        else
+        {
+            // дальняя атака
+            stream.Write((short)(ComputedDelay / 10 + 10)); // TODO  +10 It became visible flying arrows 
+            stream.Write((short)(_skill.Template.ChannelingTime / 10 + 10));
+            stream.Write((byte)0); // f - When changed to 1 when firing an auto-casting skill, will make the little blue arrow.
+            stream.WritePisc(_id, _skill.Template.FireAnim?.Id ?? 0); // added skill type here in 3.0.3.0
+            stream.Write((byte)0); // flag
+        }
+    }
 }
diff --git a/AAEmu.Game/Models/Game/Skills/Effects/SpecialEffects/AutoAttack.cs b/AAEmu.Game/Models/Game/Skills/Effects/SpecialEffects/AutoAttack.cs
index 6c6a67d6ca..24e1711c70 100644
--- a/AAEmu.Game/Models/Game/Skills/Effects/SpecialEffects/AutoAttack.cs
+++ b/AAEmu.Game/Models/Game/Skills/Effects/SpecialEffects/AutoAttack.cs
@@ -22,5 +22,7 @@ public override void Execute(BaseUnit caster,
     {
         // TODO start auto attack...
         if (caster is Character) { Logger.Debug("Special effects: AutoAttack value1 {0}, value2 {1}, value3 {2}, value4 {3}", value1, value2, value3, value4); }
+
+        // skill_effect_Id->effect_id->skill_Id автоатаки 1->1->2, 2->2->3, 51->36->4
     }
 }
diff --git a/AAEmu.Game/Models/Game/Skills/Skill.cs b/AAEmu.Game/Models/Game/Skills/Skill.cs
index 4ce81163ca..75f8a988bb 100644
--- a/AAEmu.Game/Models/Game/Skills/Skill.cs
+++ b/AAEmu.Game/Models/Game/Skills/Skill.cs
@@ -552,6 +552,14 @@ public void Cast(BaseUnit caster, SkillCaster casterCaster, BaseUnit target, Ski
     {
         if (caster is not Unit unit) { return; }
 
+        var delay = 300;
+        // Will delay for 300 Milliseconds to eliminate the hanging of the skill
+        if (!caster.CheckInterval(delay))
+        {
+            //Logger.Trace($"Skill: CooldownTime [{delay}]!");
+            return;
+        }
+
         if (!_bypassGcd)
         {
             var gcd = Template.CustomGcd;
@@ -731,7 +739,7 @@ public void StartChanneling(BaseUnit caster, SkillCaster casterCaster, BaseUnit
             doodad.Spawn();
         }
 
-        caster.BroadcastPacket(new SCSkillFiredPacket(Id, TlId, casterCaster, targetCaster, this, skillObject), true);
+        caster.BroadcastPacket(new SCSkillFiredPacket(Id, TlId, casterCaster, targetCaster, this, skillObject, caster), true);
         unit.SkillTask = new EndChannelingTask(this, caster, casterCaster, target, targetCaster, skillObject, doodad);
         TaskManager.Instance.Schedule(unit.SkillTask, TimeSpan.FromMilliseconds(Template.ChannelingTime));
     }
@@ -780,7 +788,7 @@ public void ScheduleEffects(BaseUnit caster, SkillCaster casterCaster, BaseUnit
         if (Template.FireAnim != null && Template.UseAnimTime)
             totalDelay += (int)(Template.FireAnim.CombatSyncTime * (unit.GlobalCooldownMul / 100));
 
-        caster.BroadcastPacket(new SCSkillFiredPacket(Id, TlId, casterCaster, targetCaster, this, skillObject)
+        caster.BroadcastPacket(new SCSkillFiredPacket(Id, TlId, casterCaster, targetCaster, this, skillObject, caster)
         {
             ComputedDelay = (short)totalDelay
         }, true);
